### DISCUSSION TITLE
[💡]: Coverage for spawn processes

### DISCUSSION THREAD

**Discussion by cv65kr** [roadrunner-server/roadrunner#1440] [Q&A] [answered]
### Plugin

None

### I have an idea!

There is no possibility to calculate coverage for integration tests.

To write integration test we need to run roadrunner via phpunit e.g. for Temporal plugin - https://github.com/temporalio/sdk-php/tree/master/testing#quick-start, but the same can be happen for GRPC plugin.

So when we run test PS output looks like that: phpunit -> roadrunner -> workers.

```
root      6555     0  0 15:58 pts/0    00:00:00 bash
root     25163  6555 35 17:45 pts/0    00:00:08  \_ php vendor/bin/phpunit -d memory_limit=512M --no-interaction --debug --verbose --configuration phpunit.xml.dist --coverage-html=coverage-report-phpunit.html --group e2e
root     25168 25163  1 17:45 pts/0    00:00:00      \_ ./temporal-test-server 7233 --enable-time-skipping
root     25173 25163  1 17:45 pts/0    00:00:00      \_ /usr/local/bin/rr serve -c .rr.test.yaml -w tests
root     25184 25173 10 17:45 pts/0    00:00:02          \_ php /src/bin/console app:temporal-worker -p 10 --quiet --no-ansi --no-interaction --env=test --task-queue default
root     25197 25173  8 17:45 pts/0    00:00:01          \_ php /src/bin/console app:temporal-worker -p 10 --quiet --no-ansi --no-interaction --env=test --task-queue default
```

I tried pcov and xdebug and result is the same - code is not covered.

Generally I don't have idea how solve this problem, so I will raise it here to open discussion.
Maybe new plugin which generate coverage files for spawned workers? Because it's relatively easy to merge coverage files. Maybe new phpunit driver?

**Comment by rustatian** [comment]
Hey @cv65kr 👋🏻 
I transferred the ticket into the discussion.

> Maybe new plugin which generate coverage files for spawned workers?

I guess this is a good idea, but this is hard for me to help here since I never run a `phpunit` and have no clue how this thing works 😭 

Might be @butschster / @wolfy-j might give the advice here 💯

**Comment by acelaya** [comment]
I would also be interested in supporting some way to collect code coverage in RoadRunner, for integration/E2E tests.

I have only tried using `pcov` https://github.com/sebastianbergmann/php-code-coverage/issues/1003, but I understand by the original post that xdebug wouldn't work either

**Reply by rustatian** [reply]
I'm open to suggestions. Anything needed from the RR side I'm ready to implement, but I can't help with the PHP part since I don't know PHP at all.

**Reply by acelaya** [reply]
I don't know how RoadRunner internals work, so I might be wrong, but I would say no PHP is needed per se.

Basically, we need a way to collect code coverage in the context of an app being served with RoadRunner, either by making existing mechanisms work (xdebug or pcov, which are native PHP extensions), or by implementing a new way to do it which is embedded in RoadRunner.

The first option would be more convenient, as existing frameworks (like PHPUnit) already know how to consume them, but it would always be possible to implement new drivers if needed.

Alternatively, if RoadRunner was able to generate code coverage reports in some standard format by itself, there're also tools to parse, convert and/or merge them.

Clover is a well known format used by many programming languages, for example.

**Reply by rustatian** [reply]
> I don't know how RoadRunner internals work

RR in general is a process manager. It has nothing specific to the PHP inside.

> Basically, we need a way to collect code coverage in the context of an app being served with RoadRunner, either by making existing mechanisms work (xdebug or pcov, which are native PHP extensions), or by implementing a new way to do it which is embedded in RoadRunner.

As far as I know, `xdebug` works pretty good, here is the sample configurations for it: https://roadrunner.dev/docs/php-debugging/current/en

I don't know these frameworks for generating codecov for PHP (like pcov), but could be as a solution, RR could start some RPC endpoint (not a plugin, but bundled inside), receive code coverage reports and send them to the PHPUnit?

**Reply by acelaya** [reply]
> I don't know these frameworks for generating codecov for PHP (like pcov), but could be as a solution, RR could start some RPC endpoint (not a plugin, but bundled inside), receive code coverage reports and send them to the PHPUnit?

I think something like this might definitely be an improvement, but I would focus on the simple use case for now.

For example, this code snippet using pcov to collect code:

```php
use function pcov\start;
use function pcov\stop;
use function pcov\waiting;

start();

// Use some code here, like invoking functions, etc

stop();
var_dump(waiting());
```

If you run that directly in the console (as in `php myscript.php`), this var_dump will print some stuff.

If you run it using php-fpm or other servers like openswoole, it will also print some stuff.

However, if that snippet is run inside a request handled with RoadRunner, it prints nothing, so something is "preventing" pcov from collecting the source code for code coverage.

> As far as I know, xdebug works pretty good, here is the sample configurations for it

I will try later today the snippet above with xdebug, as I haven't tried that. Based on the original comment, I assumed it wouldn't work.

**Reply by rustatian** [reply]
Ah, gotcha. I guess `pcov` is mistakenly using `stdout` to print its data. Wrongly, because generally all logs, all informational stuff should be sent to `stderr`.

Is it possible to redirect (or use some `pcov` option) to send the data through `stderr`?

**Reply by acelaya** [reply]
Sorry, maybe I used the wrong wording 😅 

pcov does not "print" anything. The `waiting` function returns an array of files that have been "included" between `start` and `stop` calls. It also has other helpers that return covered lines and such.

Basically what I meant is that in the snippet above `waiting` returns an empty array when in the context of a RoadRunner request, and in the rest of cases it returns the expected result.

**Reply by rustatian** [reply]
> Sorry, maybe I used the wrong wording

Ahah, no problems at all 😃

> pcov does not "print" anything. The waiting function returns an array of files that have been "included" between start and stop calls. It also has other helpers that return covered lines and such.

Hm, this is strange. Well, the case is that this is just a general PHP CLI process. It is the same if you run `php myscript.php`. 

Could you please send me a `worker.php` (code example with `pcov` in RR worker format) + `composer.json`? That way I'll be able to investigate where the missing output is.

**Reply by acelaya** [reply]
> Could you please send me a worker.php (code example with pcov in RR worker format) + composer.json? That way I'll be able to investigate where the missing output is.

Definitely. I will try to put something together this afternoon, and ideally test with xdebug as well.

**Reply by acelaya** [reply]
I was putting together a simple example, and I have found something.

It does seem to capture coverage somehow. However, in my real project, everything seems to have been autoloaded by the time `start` is called.

Here I was able to "delay" that until after the call, and then the included files appear are returned by `waiting`, as expected.

I will investigate a bit further in this direction.

**Reply by acelaya** [reply]
Hi!

I had to park this a few months ago, but I'm back at investigating why I could not get code coverage generated with `pcov` extension in my main project, but it worked on a simple example I created.

It turns out the problem is on the `pcov.directory` setting. This value resolves to `src` by default, if that directory exists, which turned out to be the case in the example project.

However, my main project does not have an `src` directory in the root folder, because it's a multi-module application.

I had `pcov.directory=module` set in a custom `php.ini` file, which works fine when running PHP scripts with other means, but roadrunner seems to be ignoring this. I have also tried `ini_set('pcov.directory', 'module');` and even `ini_set('pcov.directory', '/actual/absolute/path/to/module');` with no success.

If I print the result of `ini_get('pcov.directory')`, it always returns an empty value, and that's what's messing the code coverage.

I'll continue updating here once I know more.

**Reply by acelaya** [reply]
Passing ` -dpcov.directory=module` to the server worker command doesn't seem to work either.

**Reply by rustatian** [reply]
Hey @acelaya 👋 Thanks for working on that 👍. If you need any help from me, feel free to ping me on our Discord or here 😃

**Reply by acelaya** [reply]
Ok, I finally found the root cause. It was related with `pcov.directory`, but not exactly.

The problem seems to be somehow related with the fact that I have my .rr.yml config files in a subdirectory (to keep things tidy, as the project root is already full of other config files), which affects some relative paths defined in it, like `server.command`.

If I move that config file to the project root directory and adjust those relative paths, then code coverage with pcov works as expected, without changing anything else.

**Reply by rustatian** [reply]
Great job 👍 
Could you please write here a small instruction what to do in such a case? I mean, when a user needs to generate code coverage. Where should be `.rr.yaml`, what should be in the config, might be how to start RR (might be use `-w` to set a working directory).
I'll put this in the documentation.

**Reply by acelaya** [reply, marked as answer]
I actually didn't think about the `-w` flag!

I suppose the working directory is inferred from the rr.yaml config file location, if not specified. What I did was moving that file to the project root.

However, I have just tried moving the file to its original subfolder, and specified the working directory via `-w`, and that also worked, with the only implication that routes inside the rr.yaml file needed to be adjusted to be relative to the project root, not the file location.

So I guess you can document those two approaches.

**Reply by rustatian** [reply]
Yeah, definitely :) But, I'm not a PHP dev (I mean, at all) 😆 
It would be super cool if you'd be able to write a short instruction here about I wrote previously. And I'll put it in the proper docs section 😃

**Reply by acelaya** [reply]
I guess the explanation depends on what is intended to be documented. When I first started participating in this discussion, I was having some problems generating code coverage for E2E tests, where the test runner is executed on a different process than the code under test.

Before using RoadRunner, I was using a different app server (openswoole), and for that one, I came up with a workaround which required to conditionally and programmatically call the appropriate logic to collect and generate code coverage reports, which works out of the box for unit tests where the code under tests is imported in the same process executed by the test runner.

This logic is a bit cumbersome, and has some implications, like the fact that you need to make sure there's only one worker serving HTTP requests, otherwise you end-up again with multiple processes loading your source code, and you would need to coordinate them somehow to properly generate the coverage reports. Also, logic running in job workers would never be marked as covered.

When I moved to RoadRunner, I assumed this approach would work as well, but my surprise was that I was getting reports with 0% of coverage.

After googling a bit I ended up here, and incorrectly assumed there was an actual known issue with RoadRunner and coverage reporting tools, but it turns out the original author of this discussion, was actually suggesting something else, implementing something in RoadRunner itself, so that the server workers could report code coverage and let the test runner on a different process consume it and generate coverage reports transparently as you would do with unit tests.

This approach is obviously much better, but considering I already had the other one working, I thought it would be easier to get that one working in the short term. Also, it has the benefit of being agnostic to the app server.

With the context clear, it turned out the reason I was getting a 0% coverage was because of the incorrect setting of the working directory, which could be solved with any of the two approaches described here https://github.com/orgs/roadrunner-server/discussions/1440#discussioncomment-8486432

~It's also worth mentioning I only tried with `pcov` (one of the two coverage drivers supported by `PHPUnit`, the most popular PHP test runner/test framework). `XDebug`, the other coverage driver, might behave differently.~ EDIT: I'm now using xdebug and works the same.

**Reply by rustatian** [reply]
Cool, thanks for the explanation 👍

**Comment by cv65kr** [comment]
Long story short:

If you want to have a integration test, on the beginning of test you need to run roadrunner (and this is separate process).
So now your top looks like that:
- phpunit (Main process)
- roadrunner (2 workers)
- - php worker
- - php worker.

According to way how coverage is calculated in phpunit - https://github.com/sebastianbergmann/php-code-coverage/blob/main/src/Driver/PcovDriver.php the test block looks like
```
START test
- pcov/xdebug START
- detached RR
- test
- assertions
- pcov/xdebug START
END test
```

So in general main process in this case phpunit is not able to calculate and sum coverage from detached php workers.

Actual behaviour: coverage is calculated without php workers


---
So @acelaya the idea to add pcov start/end in worker is good start, but it's kinda bad solution because you need to have production and testing workers - so you need to maintain two things.

The problem is still the same, phpunit as a main process still is not able to calculate coverage from detached processes.

```
Ah, gotcha. I guess pcov is mistakenly using stdout to print its data. Wrongly, because generally all logs, all informational stuff should be sent to stderr.

Is it possible to redirect (or use some pcov option) to send the data through stderr?
```
pcov/xdebug functions not printing anything to stdout:
https://github.com/sebastianbergmann/php-code-coverage/blob/main/src/Driver/XdebugDriver.php#L110
https://github.com/sebastianbergmann/php-code-coverage/blob/main/src/Driver/PcovDriver.php#L50
https://github.com/krakjoe/pcov

**Reply by rustatian** [reply]
Do you have any suggestions on how we can resolve this issue?

**Reply by cv65kr** [reply]
I was thought about something like that: 
1. pcov/xdebug can send output of coverage function via RPC to RR from all php workers
2. RR aggregate result for all php workers and store in memory
3. Create new driver for phpunit e.g. [RoadrunnerDriver](https://github.com/sebastianbergmann/php-code-coverage/tree/main/src/Driver) which can get result from point 2.

The problem was that I was not able to find elegant solution to keep prod/testing worker as the same.

Since I didn't have time I don't put much attention here, but I will think about that.

**Reply by rustatian** [reply]
> Create new driver for phpunit e.g. [RoadrunnerDriver](https://github.com/sebastianbergmann/php-code-coverage/tree/main/src/Driver) which can get result from point 2.

Looks fairly easy to implement, not for my 0 PHP knowledge for sure 😄 

> The problem was that I was not able to find elegant solution to keep prod/testing worker as the same.

Yeah, this is the case. I guess, one solution is to have a separate configurations or might be use [includes](https://github.com/roadrunner-server/roadrunner/issues/935).

**Reply by rustatian** [reply]
So, if someone would be willing to implement this feature from PHP side, I'm all yours to help from RR side ⚡

**Reply by cv65kr** [reply]
I can try to make some POC, I see couple challenges here but even if fail we will be have some picture.

> Yeah, this is the case. I guess, one solution is to have a separate configurations or might be use [includes](https://github.com/roadrunner-server/roadrunner/issues/935).

Problem is a bit different, I would like avoid situation where we need to have two .rr.yaml:
.rr.prod.yaml - worker-prod.php
.rr.test.yaml - worker-test.php

and one difference between those files is 
```
<?php

xdebug_start(); // this

// worker code

xdebug_stop(); // and this

?>
```

**Reply by rustatian** [reply]
Yeah, looks weird 😞 
So, theoretically, with the RR phpunit driver we have the ability to avoid this case, right?

**Reply by cv65kr** [reply]
> pcov/xdebug can send output of coverage function via RPC to RR from all php workers

By this point I had in head that we need to have 2 files prod/test 😢 

The problem is that I didn't saw auto measuring when script starts and output send to stdout. I will check documentation once again.

**Reply by acelaya** [reply]
@cv65kr I was not trying to say my example above is the way to go. With that I just wanted to highlight that pcov is currently not reporting any coverage, regardless of running on the same process.

Then there's the fact about PHPUnit running on a different process than the code itself, but that works on the assumption collecting coverage works, which is not the case.

I already had E2E tests with code coverage working, but using another process manager: you can see it here https://github.com/sebastianbergmann/php-code-coverage/issues/1003

What you see there works with any server, except RoadRunner. It's ugly, and I would rather have what you propose, but that won't work at this point.

I will try to put the minimal example together, and see if it also doesn't work with xdebug, so that @rustatian can investigate.

Next steps would be working on all the orchestration you suggest.

**Reply by cv65kr** [reply]
I checked as fast poc - storing all coverage outputs (from php workers) to files and merge it in the main process. Generally looks good.

Like I said the ugly thing is that we need to keep two workers files for testing/prod, but this solution is possible to do even without RR.

I will think about single entrypoint and will try to back to you asap.

**Reply by wolfy-j** [reply]
May I ask why not calculate coverage within tests without the use of RR? All drivers (HTTP, GRPC and Queue) are properly isolated from RR instance, so calling GRPC service via stub server will fully emulate the RR behaviour. You can even reuse it between tests to check for memory leaks.

**Reply by wolfy-j** [reply]
We don't have stub server for queues (we can build one if really needed), but for GRPC you can test using Invoker - https://github.com/roadrunner-php/grpc/blob/3.x/src/Invoker.php

From PHP standpoint there are no other magic going on, it can be executed in a separate CLI and should produce response similar to RR.

**Reply by cv65kr** [reply]
What about temporal sdk? which uses temporal test server and you are not able to isolate it

**Reply by wolfy-j** [reply]
Yes, Temporal will be an exception, unlike all other drivers it uses async pipeline which is almost impossible to decouple from code (without reinventing temporal in PHP). There a internal trick we can use to replay history of communication between RR and worker, but it's not exposed to user level. So for Temporal you might still need to use external coverage instrument, I have not worked with them in such regard so it's hard for me to advice. I can only say that RR processes behave similar to simple PHP CLI, so if you can do coverage of CLI you can do coverage of any RR worker.
